Fix #76093: Format strings w/o loss of precision w/ FORMAT_TYPE_DECIMAL

lucaswerkmeister opened the pull request:
Resubmit of #9909, which was itself a resubmit of #7782. Rebased onto latest master. [Request #76093](https://bugs.php.net/bug.php?id=76093) is classified as a feature/change request rather than a bug, which I think I agree with, so IIUC this is supposed to be based on `master` rather than `PHP-8.0`.

lucaswerkmeister commented:
Hm, there’s a CI error I’m not sure I understand :/

<details>
<summary>runtime error: 1e+19 is outside the range of representable values of type 'long' – UndefinedBehaviorSanitizer</summary>

```
========DIFF========
--
       ["currency"]=>
       string(29) "$9,999,999,999,999,999,999.00"
     }
029+ /__w/php-src/php-src/ext/intl/formatter/formatter_format.c:105:13: runtime error: 1e+19 is outside the range of representable values of type 'long'
030+ SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /__w/php-src/php-src/ext/intl/formatter/formatter_format.c:105:13 in 
     array(6) {
       ["input"]=>
       float(1.0E+19)
--
========DONE========
FAIL Bug #76093 (NumberFormatter::format loses precision) [ext/intl/tests/bug76093.phpt] 
```
</details>

If I understand correctly, the UB is in the `if` branch of the following part of the diff:

```diff
@@ -79 +103,7 @@ PHP_FUNCTION( numfmt_format )
-			int64_t value = (Z_TYPE_P(number) == IS_DOUBLE)?(int64_t)Z_DVAL_P(number):Z_LVAL_P(number);
+			int64_t value;
+			if (Z_TYPE_P(number) == IS_DOUBLE) {
+				value = (int64_t)Z_DVAL_P(number);
+			} else {
+				convert_to_long(number);
+				value = Z_LVAL_P(number);
+			}
```

But according to that diff, the cast already existed (in a ternary instead of `if`/`else`). So is the error only showing up because this previously wasn’t hit by a test?

lucaswerkmeister commented:
Hm, I think this comment in the test is referring to the same thing, actually:

```php
# Also, casting from double to int64 when the int64 range
# is exceeded results in an implementation-defined value.
```

Girgias commented on ext/intl/formatter/formatter_format.c:
```diff
@@ -76,7 +100,13 @@ PHP_FUNCTION( numfmt_format )
+			if (Z_TYPE_P(number) == IS_DOUBLE) {
+				value = (int64_t)Z_DVAL_P(number);
+			} else {
+				convert_to_long(number);
+				value = Z_LVAL_P(number);
```
This will cast non-numeric strings to 0, which is not what should happen.

Same as the other cases using ``convert_to_long()`` or ``convert_to_double()``

lucaswerkmeister replied:
Okay, what should happen instead?

lucaswerkmeister replied:
Or to put the question slightly differently: I’m not very familiar with the php-src codebase, and couldn’t even find documentation on what exactly `convert_to_long()` does – if there’s another paradigm for what the code should do with non-numeric strings, such as throwing an error, can you point me to another place in PHP that uses this paradigm, so I can see how it’s implemented there?

lucaswerkmeister replied:
I switched it to `zval_try_get_long()` (in a separate commit), but it seems to introduce a lot of deprecation warnings that are currently causing tests to fail. Should I add the deprecations to the expected test output, or should the deprecations not happen after all?

Also added a `throw` for cases where the `INT32` type is used with numbers that exceed 32 bits of precision, but I’m not sure whether that’s what you had in mind or not. Happy to change this part.

`convert_to_double()` not converted, because I couldn’t find a `zval_try_get_double()` or equivalent function.

Girgias commented on ext/intl/formatter/formatter_format.c:
```diff
@@ -135,11 +184,20 @@ PHP_FUNCTION( numfmt_format_currency )
+
 	/* Parse parameters. */
-	if( zend_parse_method_parameters( ZEND_NUM_ARGS(), getThis(), "Ods",
-		&object, NumberFormatter_ce_ptr,  &number, &currency, &currency_len ) == FAILURE )
-	{
-		RETURN_THROWS();
+	if (object) {
+		ZEND_PARSE_PARAMETERS_START(2, 2)
+			Z_PARAM_STR_OR_NUMBER(number)
+			Z_PARAM_STRING(currency, currency_len)
+		ZEND_PARSE_PARAMETERS_END();
+	} else {
+		ZEND_PARSE_PARAMETERS_START(3, 3)
+			Z_PARAM_OBJECT_OF_CLASS(object, NumberFormatter_ce_ptr)
+			Z_PARAM_STR_OR_NUMBER(number)
+			Z_PARAM_STRING(currency, currency_len)
+		ZEND_PARSE_PARAMETERS_END();
 	}
```
Why does the currency method/function support strings here? That seems to make no sense.

lucaswerkmeister replied:
It’s already in a separate commit, so I could remove it again, but I don’t see why it doesn’t make sense? Currency amounts can exceed int/float precision just like other numbers can…

Girgias replied:
This function is to deal with the currency formatter, so using the decimal formatter somewhere under the hood is not what I expect.

Also I highly doubt one can have a currency amount of 18 quintillion (/trillion) overflow, even if that is in cents.

lucaswerkmeister replied:
Alright, currency part removed.

mvorisek commented on ext/intl/formatter/formatter_format.c:
```diff
@@ -53,10 +64,23 @@ PHP_FUNCTION( numfmt_format )
+				type = FORMAT_TYPE_DECIMAL;
+				break;
+			default:
+				zend_argument_type_error(1, "must be of type int|float|string, %s given", zend_zval_type_name(number));
```
with `Z_PARAM_STR_OR_NUMBER`, can this line be reached?

lucaswerkmeister replied:
Hm, good point, I think it can’t be – at least I can’t seem to reach it with some manual testing. (All the error messages I get name the parameter type as `string|int|float`, not `int|float|string` as in this line, so those errors presumably come from `Z_PARAM_STR_OR_NUMBER` instead.)

I seem to have introduced this in #9909, possibly misunderstanding [this comment](https://github.com/php/php-src/pull/7782#discussion_r886547876) despite otherwise not going back to `zend_parse_method_parameters()` after all. I’ll change it back and see if any CI complains.

Girgias replied:
I do think the approach highlighted by @cmb69 to use a generic ``z`` to be potentially more sensible.

lucaswerkmeister replied:
Done. (I think I misunderstood the suggestion before.) Makes the commit smaller, at least, which is nice.

Girgias commented on ext/intl/formatter/formatter_format.c:
```diff
@@ -103,6 +131,24 @@ PHP_FUNCTION( numfmt_format )
+			break;
+
 		case FORMAT_TYPE_CURRENCY:
 			if (getThis()) {
```
Aside Nit: this could now be ``if (object)``

lucaswerkmeister replied:
I tried it, but it caused several test failures (seemingly `object` was still null when I wouldn’t have expected it to be).

Girgias commented on ext/intl/formatter/formatter_format.c:
```diff
@@ -103,6 +131,24 @@ PHP_FUNCTION( numfmt_format )
 			break;
+
+		case FORMAT_TYPE_DECIMAL:
+			if (!try_convert_to_string(number)) {
+				RETURN_THROWS();
+			}
```
This _could_ be simplified as at this point the value should be ``int|float|string`` and integers and floats always have string representations.

Girgias commented on ext/intl/tests/bug48227.phpt:
```diff
@@ -5,20 +5,50 @@ intl
     }
 }
 
+echo("\nNon-OOP\n");
```
```suggestion
echo("\nProcedural\n");
```

lucaswerkmeister replied:
Done.

Girgias commented on ext/intl/tests/bug76093.phpt:
```diff
@@ -0,0 +1,81 @@
+--TEST--
+Bug #76093 (NumberFormatter::format loses precision)
+--SKIPIF--
+<?php if( !extension_loaded( 'intl' ) ) print 'skip'; ?>
```
Please use the EXTENSIONS section, if other tests in ext/intl don't use it they need to be converted and backported (as a separate PR).

lucaswerkmeister replied:
Done.

Girgias reviewed (changes requested):
I'm not sure adding a ZPP modifier for one (/two) case is totally worth it, as the logic inside of it doesn't even really utilize the benefit of it as values may still be cast around within the function body.

In any case, Stringable objects need to be handle properly.

lucaswerkmeister commented on Zend/zend_API.h:
```diff
@@ -2506,6 +2519,18 @@ static zend_always_inline bool zend_parse_arg_str_or_long(zval *arg, zend_string
+		return zend_parse_arg_number_slow(arg, dest, arg_num);
+	}
+	return 1;
+}
```
Ack, still TODO. I’ll see if I can figure it out, but wanted to push a new version for some of the other comments first.

lucaswerkmeister commented:
Rebased and updated – reverted the argument parsing to `zend_parse_method_parameters()` style (with `z` instead of `n`, and manual `zend_argument_type_error()` for unexpected argument type), and used `zval_try_get_long()`. Stringable handling still TBD.

Girgias commented on ext/intl/tests/bug48227.phpt:
```diff
@@ -5,20 +5,50 @@ intl
 
+$testcases = ['', 1, NULL, true, false, [], (object)[]];
+
+echo("OOP\n");
```
```suggestion
echo "OOP\n";
```
``echo`` is not a function, please do not use parenthesis.

Girgias commented on ext/intl/tests/bug48227.phpt:
```diff
@@ -5,20 +5,50 @@ intl
     }
 }
 
+echo("\nProcedural\n");
```
Ditto

Girgias commented on ext/intl/formatter/formatter_format.c:
```diff
@@ -53,30 +53,55 @@ PHP_FUNCTION( numfmt_format )
+			case IS_STRING:
+				type = FORMAT_TYPE_DECIMAL;
+				break;
+			default:
+				zend_argument_type_error(1, "must be of type int|float|string, %s given", zend_zval_type_name(number));
```
This will fail with a Stringable object. Also, booleans are now not accepted in weak mode (they behave like ``int``) but that is more of a whatever, so just add ``IS_OBJECT`` at that same level as ``IS_STRING`` and the type error caused by the conversion to string will handle the bailing out.

lucaswerkmeister replied:
Alright, Stringable objects should now work with `TYPE_DEFAULT` / `TYPE_DECIMAL` (and I added it to the test) – but not with `TYPE_INT64`: it looks like `zval_try_get_long()` doesn’t support Stringable objects (whereas `try_convert_to_string()` does).

Girgias replied:
This is actually expected behaviour, stringable objects do not coerce to int, float, or bool.

lucaswerkmeister replied:
Alright :+1:

Girgias commented on ext/intl/formatter/formatter_format.c:
```diff
@@ -53,30 +53,55 @@ PHP_FUNCTION( numfmt_format )
+		{
+			bool failed = true;
+			int64_t value_64 = zval_try_get_long(number, &failed);
+			if (failed || value_64 < -2147483648 || value_64 > 2147483647) {
+				zend_argument_value_error(object ? 1 : 2, "must be numeric and fit in 32 bits");
+				RETURN_THROWS();
+			}
```
The ``failed`` case should be split into a different error, as it is a type error.

Girgias commented on ext/intl/formatter/formatter_format.c:
```diff
@@ -53,30 +53,55 @@ PHP_FUNCTION( numfmt_format )
+			bool failed = true;
+			int64_t value = zval_try_get_long(number, &failed);
+			if (failed) {
+				zend_argument_value_error(object ? 1 : 2, "must be numeric");
```
```suggestion
				zend_argument_type_error(object ? 1 : 2, "must be of type int when argument #2 ($format) is NumberFormatter::TYPE_INT64");
```

Girgias replied:
Ah yes, well the argument number can be a sprintf placeholder
``zend_argument_type_error(object ? 1 : 2, "must be of type int when argument #%d ($format) is NumberFormatter::TYPE_INT64", object ? 2 : 3);

lucaswerkmeister replied:
Got it, thanks!
